File: apps/api/app/schemas/inspo.py
"""Pydantic models for Inspo Boards endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── Request models ────────────────────────────────────────


class InspoBoardCreate(BaseModel):
    """POST /inspo/boards — create a new board."""
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    brand_id: Optional[str] = None


class InspoBoardUpdate(BaseModel):
    """PATCH /inspo/boards/{id} — update board name/description."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=2000)


class InspoItemCreate(BaseModel):
    """POST /inspo/boards/{board_id}/items — add an item to a board."""
    content_type: str = Field(default="text", pattern="^(text|link|image|video|voice_note)$")
    title: str = Field(default="", max_length=500)
    content_text: str = Field(default="", max_length=500000)
    source_url: Optional[str] = None
    source_tag: str = Field(
        default="",
        max_length=500,
        description="Where this came from (e.g. 'Alex Hormozi, YouTube'). Auto-populated for links.",
    )
    intent_note: str = Field(
        default="",
        max_length=5000,
        description="Why you saved it and what the AI should derive from it.",
    )
    tags: List[str] = Field(default_factory=list)
    is_starred: bool = False
    metadata: Dict[str, Any] = Field(default_factory=dict)


class InspoItemUpdate(BaseModel):
    """PATCH /inspo/items/{id} — update an item."""
    title: Optional[str] = Field(None, max_length=500)
    content_text: Optional[str] = Field(None, max_length=500000)
    source_tag: Optional[str] = Field(None, max_length=500)
    intent_note: Optional[str] = Field(None, max_length=5000)
    tags: Optional[List[str]] = None
    is_starred: Optional[bool] = None
    sort_order: Optional[int] = None


# ── Response models ───────────────────────────────────────


class InspoBoardSummary(BaseModel):
    """Returned in list view (GET /inspo/boards)."""
    id: str
    name: str
    description: str = ""
    brand_id: Optional[str] = None
    item_count: int = 0
    starred_count: int = 0
    created_at: datetime
    updated_at: datetime


class InspoItemSummary(BaseModel):
    """An item within a board."""
    id: str
    board_id: str
    content_type: str
    title: str = ""
    content_text: str = ""
    source_url: Optional[str] = None
    source_tag: str = ""
    intent_note: str = ""
    media_path: Optional[str] = None
    tags: List[str] = Field(default_factory=list)
    is_starred: bool = False
    metadata: Dict[str, Any] = Field(default_factory=dict)
    sort_order: int = 0
    created_at: datetime
    updated_at: datetime


class InspoBoardDetail(BaseModel):
    """Returned in detail view (GET /inspo/boards/{id})."""
    id: str
    name: str
    description: str = ""
    brand_id: Optional[str] = None
    items: List[InspoItemSummary] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class InspoItemCreated(BaseModel):
    """Returned after POST /inspo/boards/{board_id}/items."""
    id: str
    content_type: str
    title: str
    source_tag: str = ""
    message: str = "Item added to board"


# Alias used by the router for full item responses
InspoItemDetail = InspoItemSummary

File: apps/web/src/lib/api/inspo.ts
/**
 * Inspo Boards API -- boards and items CRUD.
 */

import { apiFetch } from "./client";

// ── Types ────────────────────────────────────────────────

export interface InspoBoardSummary {
  id: string;
  user_id: string;
  brand_id: string | null;
  name: string;
  description: string | null;
  item_count: number;
  created_at: string;
  updated_at: string;
}

export interface InspoItemDetail {
  id: string;
  board_id: string;
  user_id: string;
  content_type: "text" | "link" | "image" | "video" | "voice_note";
  content_text: string | null;
  source_url: string | null;
  source_tag: string | null;
  intent_note: string | null;
  media_path: string | null;
  tags: string[];
  is_starred: boolean;
  metadata: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface InspoBoardDetail extends InspoBoardSummary {
  items: InspoItemDetail[];
}

// ── API Methods ──────────────────────────────────────────

export const inspoApi = {
  listBoards: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<InspoBoardSummary[]>(`/inspo/boards${qs}`);
  },

  createBoard: (data: {
    name: string;
    description?: string;
    brand_id?: string;
  }) =>
    apiFetch<InspoBoardSummary>("/inspo/boards", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  getBoard: (boardId: string) =>
    apiFetch<InspoBoardDetail>(`/inspo/boards/${boardId}`),

  updateBoard: (
    boardId: string,
    data: { name?: string; description?: string; brand_id?: string }
  ) =>
    apiFetch<InspoBoardSummary>(`/inspo/boards/${boardId}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  deleteBoard: (boardId: string) =>
    apiFetch<void>(`/inspo/boards/${boardId}`, { method: "DELETE" }),

  listItems: (boardId: string, starredOnly?: boolean, tag?: string) => {
    const params = new URLSearchParams();
    if (starredOnly) params.set("starred_only", "true");
    if (tag) params.set("tag", tag);
    const qs = params.toString();
    return apiFetch<InspoItemDetail[]>(
      `/inspo/boards/${boardId}/items${qs ? `?${qs}` : ""}`
    );
  },

  createItem: (
    boardId: string,
    data: {
      content_type: string;
      content_text?: string;
      source_url?: string;
      source_tag?: string;
      intent_note?: string;
      tags?: string[];
      is_starred?: boolean;
      metadata?: Record<string, any>;
    }
  ) =>
    apiFetch<InspoItemDetail>(`/inspo/boards/${boardId}/items`, {
      method: "POST",
      body: JSON.stringify(data),
    }),

  getItem: (itemId: string) =>
    apiFetch<InspoItemDetail>(`/inspo/items/${itemId}`),

  updateItem: (
    itemId: string,
    data: {
      content_type?: string;
      content_text?: string;
      source_url?: string;
      source_tag?: string;
      intent_note?: string;
      tags?: string[];
      is_starred?: boolean;
      metadata?: Record<string, any>;
    }
  ) =>
    apiFetch<InspoItemDetail>(`/inspo/items/${itemId}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  deleteItem: (itemId: string) =>
    apiFetch<void>(`/inspo/items/${itemId}`, { method: "DELETE" }),

  toggleStar: (itemId: string) =>
    apiFetch<InspoItemDetail>(`/inspo/items/${itemId}/star`, {
      method: "PATCH",
    }),
};
